phoenix-core/src/it/java/org/apache/phoenix/end2end/BackwardCompatibilityIT.java

```diff
@@ -0,0 +1,396 @@
+        executeQueryWithClientVersion(compatibleClientVersion, QUERY);
+        assertTrue(compareOutput(ADD_DATA, QUERY));
+    }
+
+    /**
+     * Scenario: 
+     * 1. New Client connects to the updated server 
+     * 2. New Client creates tables and inserts data 
+     * 3. Old Client reads the data inserted by the old client 
+     * 4. Old Client inserts more data into the tables created by old client 
+     * 5. New Client reads the data inserted by new client
+     * 
+     * @throws Exception
+     */
+    @Test
+    public void testSelectUpsertWithOldClient() throws Exception {
+        checkForPreConditions();
+        // Insert data with new client and read with old client
+        executeQueriesWithCurrentVersion(CREATE_ADD);
+        executeQueryWithClientVersion(compatibleClientVersion, QUERY);
+        assertTrue(compareOutput(CREATE_ADD, QUERY));
+
+        // Insert more data with old client and read with new client
+        executeQueryWithClientVersion(compatibleClientVersion, ADD_DATA);
+        executeQueriesWithCurrentVersion(QUERY);
+        assertTrue(compareOutput(ADD_DATA, QUERY));
+    }
+
+    /**
+     * Scenario: 
+     * 1. Old Client connects to the updated server 
+     * 2. Old Client creates tables and inserts data 
+     * 3. New Client reads the data inserted by the old client 
+     * 4. Old Client creates and deletes the data
+     * 
+     * @throws Exception
+     */
+    @Test
+    public void testUpsertDeleteWithOldClient() throws Exception {
+        checkForPreConditions();
+        // Insert data with old client and read with new client
+        executeQueryWithClientVersion(compatibleClientVersion, CREATE_ADD);
+        executeQueriesWithCurrentVersion(QUERY);
+        assertTrue(compareOutput(CREATE_ADD, QUERY));
+
+        // Deletes with the old client
+        executeQueryWithClientVersion(compatibleClientVersion, ADD_DELETE);
+        executeQueryWithClientVersion(compatibleClientVersion, QUERY_ADD_DELETE);
+        assertTrue(compareOutput(ADD_DELETE, QUERY_ADD_DELETE));
+    }
+
+    /**
+     * Scenario: 
+     * 1. New Client connects to the updated server 
+     * 2. New Client creates tables and inserts data 
+     * 3. Old Client reads the data inserted by the old client 
+     * 4. New Client creates and deletes the data
+     * 
+     * @throws Exception
+     */
+    @Test
+    public void testUpsertDeleteWithNewClient() throws Exception {
+        checkForPreConditions();
+        // Insert data with old client and read with new client
+        executeQueriesWithCurrentVersion(CREATE_ADD);
+        executeQueryWithClientVersion(compatibleClientVersion, QUERY);
+        assertTrue(compareOutput(CREATE_ADD, QUERY));
+
+        // Deletes with the new client
+        executeQueriesWithCurrentVersion(ADD_DELETE);
+        executeQueriesWithCurrentVersion(QUERY_ADD_DELETE);
+        assertTrue(compareOutput(ADD_DELETE, QUERY_ADD_DELETE));
+    }
+    
```

Comment:
Most backward compatibility issues we found during the 4.15 release were related to old clients unable to perform DDL operations once the cluster was upgraded or old clients unable to query views/indexes/view indexes created by a new clients. Can we add more tests like that?

Reply:
I think what you meant is creating views etc with a older client once the new client connects to the updated server. Yes, that makes sense and I have added more DDL to the add_data to have all these scenarios covered